text/0000-wallet-replay-protection.md

```diff
@@ -0,0 +1,298 @@
+
+### Separation of replay protected outputs
+
+Let's define a clear separation of outputs that are protected from those that are not. We define a protected output generator `GenP` that allows us to either `create` a protected output or `check` whether an output is `Protected`.
+
+An output is labeled as `Protected` *only if* it was created in a transaction where we contributed an input that was labeled as `Protected`. An exception to the rule are outputs that are created as a part of the _Bootstrap_ transaction discussed in the next section.
+
+#### Possible protected output generators implementations
+
+As we mentioned above, `GenP` for `Protected` outputs would need to have `create` and `check` functions defined.
+
+There are a few choices how to label outputs as `Protected` while still being able to identify them across difference devices:
+1. Call to `create` uses a new derivation path `P` that is used only for creating `Protected` outputs. Similarly `check` uses the same `P` to check whether an output is protected.
+2. Call to `create` creates a specific `r` value for `Protected` outputs e.g. they should start with `N` zeros or be divisible by some number `M`
+3. Call to `create` uses additional output information in the ~30 bytes that are available in the Bulletproofs to convey the idea whether the output is protected. Perhaps we could define a specific structure for these bytes e.g. `<scheme_version:1 byte><meta_data:4 bytes><data:25 bytes>` where `metadata` would also tell whether the `data` that follows is encrypted or not. The data could be encrypted using the `seed` key and could thus hold information on output labels which would only be available to the owner of the output. We could even include the starting bytes of the anchor that protects it if we wanted to. If we went such path, we would need to think of the possible drawbacks.
+
+In all cases, the output label should only be visible to the owner of the output. It seems necessary to have labeling information about the UTXO on the UTXO itself if we want it to be consistent with different wallet reusing the same seed. If the information is held only on the wallet side, then we hit much bigger issues because there comes a need for either a manual intervention and labeling or a robust solution for synchronization between wallets - which does not appear simple to build.
+
+### Simple bootstrapping of protected outputs
+
+We start off with all of our outputs marked as `Unprotected`. To create a `Protected` output out of nothing we create the following outputs:
+1. an `anchor` output that has a form `0*H + r*G` - generated from key derivation path `A`

```

Comment:
We should consider the anchor itself also protected. Perhaps it makes sense to say that a protected output is spendable if and only if it has nonzero value.

Reply:
This is a good point. I do however wonder whether this would close the opportunities to use 0 value outputs somewhere else. Zero value outputs are the only way to bootstrap yourself with outputs without an explicit `receive` by simply attaching them to a passing by transaction - they're the only type of outputs that can be created on either side input/output that don't require to be matched in their `v` value. My current belief is that decoys can be useful if they have a good spending pattern (similar to how people spend - whatever that looks like). Perhaps a decoy could have a 0 value input and output? I'm not sure whether what I'm saying actually holds value in practice though. But I agree that it would be nice to have all the vertices from the anchor graph be of the same type...

Reply:
We can adopt a convention that 0-valued outputs are anchors when protected and decoys when unprotected.